### PR TITLE
Introduce types `DeltaUTxO` and `DeltaWallet`

### PR DESCRIPTION
**Pull request by HeinrichApfelmus**
### Issue number

ADP-1434

### Overview

Previous work in epic ADP-1043 introduced delta encodings, DBVars, and an embedding of the wallet state and its delta encodings into a database table. It's time to integrate these tools with the wallet code. To facilitate code review, the integration proceeds in a sequence of refactorings that do not change functionality and pass all unit tests.

In this step, we finally introduce the type `DeltaUTxO` which represents a delta encoding of the `UTxO` set, i.e. a package of changes that can be applied to one `UTxO` set in order to obtain another one. The type `DeltaUTxO` is abstract, but internally, it is implemented as
```hs
data DeltaUTxO = DeltaUTxO
    { excluded :: !(Set TxIn) -- ^ First exclude these inputs
    , received :: !UTxO       -- ^ Then receive these additional outputs.
    }
```
i.e. it stores a set of `TxIn` that was spent, and a `UTxO` that was received.

Building on this, we also introduce a delta encoding `DeltaWallet` for the wallet state `Wallet s`. The main change is that the `applyBlock` function now also computes and returns this delta encoding when computing the new `Wallet s` state.

```hs
applyBlock
    :: (…)
    => Block -> Wallet s -> (FilteredBlock, (DeltaWallet s, Wallet s))
```

However, this pull request does not yet integrate this type into the `WalletState` checkpoints; this will be the subject of an upcoming pull request.

### Details

* `DeltaWallet s` currently uses the trivial delta encoding `Data.Delta.Replace` for the address discovery state.
* As indicated, the delta encodings `DeltaWallet s` are not yet used with the `Checkpoint` mechanism. Instead, they are currently discarded.
* Even though it is currently discarded, the `DeltaUTxO` type has suggested an optimization where a transaction that does not change the `UTxO` set can be discarded more quickly. This may already improve the wallet restoration benchmarks slightly.
* I have smuggled in a small cleanup commit that removes the `Dom` type class, as it is no longer used.

### Comments

* @piotr-iohk I'd be interested in the wallet restoration benchmark for this branch — I expect slight improvements for the `0.01-percent-seq` and `0.01-percent-rnd` benchmarks.

### CODE REVIEW

File: lib/core/src/Cardano/Wallet/Primitive/Types/UTxO.hs [hunk 1]
```diff
@@ -212,6 +211,52 @@ partition f (UTxO u) = bimap UTxO UTxO $ Map.partitionWithKey (const . f) u
 toList :: UTxO -> [(TxIn, TxOut)]
 toList = Map.toList . unUTxO
 
+{-------------------------------------------------------------------------------
+    Delta encodings of UTxO
+-------------------------------------------------------------------------------}
+-- | Efficient delta encoding for 'UTxO'.
+data DeltaUTxO = DeltaUTxO
+    { excluded :: !(Set TxIn) -- ^ First exclude these inputs
+    , received :: !UTxO       -- ^ Then receive these additional outputs.
```
Comment (jonathanknowles) [inline lib/core/src/Cardano/Wallet/Primitive/Types/UTxO.hs:220]: Hi @HeinrichApfelmus 

Just wondering, why not remove a level of nesting:
```suggestion
    , received :: !(Map TxIn TxOut)       -- ^ Then receive these additional outputs.
```
My (hand-waving) justification is that this makes it obvious to the reader that the type of keys in `excluded` and the type of values in `received` are in fact the same type.

(Apologies for the drive-by comment!)

File: lib/core/test/unit/Cardano/Wallet/Primitive/Types/UTxOSpec.hs [hunk 2]
```diff
@@ -50,6 +61,10 @@ spec :: Spec
 spec =
     describe "Cardano.Wallet.Primitive.Types.UTxOSpec" $ do
 
+    parallel $ describe "delta encoding" $ do
+        it "DeltaUTXO is a semigroup" $
+            property prop_deltaUTxO_semigroup
```
Comment (jonathanknowles) [inline lib/core/test/unit/Cardano/Wallet/Primitive/Types/UTxOSpec.hs]: I wonder if using `quickcheck-classes` here would be helpful?

For example: https://github.com/input-output-hk/cardano-wallet/blob/3ecad89bbcc4c4baa70e3cb4b5695aecdb0d92f5/lib/core/test/unit/Cardano/Wallet/Primitive/Types/TokenMapSpec.hs#L129

File: lib/core/src/Cardano/Wallet/Primitive/Types/UTxO.hs (hunk 1 above)
Comment (HeinrichApfelmus) [reply lib/core/src/Cardano/Wallet/Primitive/Types/UTxO.hs:220]: Well, I would argue that within the scope of the `Cardano.Wallet.Primitive.Types.UTxO` module, the fact that the key type of `UTxO` is also `TxIn` is already "in scope" — no need to put an extra highlight on it.

By using the type `UTxO`, we reuse existing concepts within the module, e.g. `excluding` in the definition of the `Semigroup` instance, or in the definition of `excludingD`.

The fact that `excludingS` and `restrictedByS` are defined in terms of the `Map` is a matter of performance, not of semantics. A semantic definition of `excludingS` would have been

    excludingS ins utxo = ins `Set.difference` dom utxo

but this definition allocates and destroys an extra `Set`.

File: lib/core/test/unit/Cardano/Wallet/Primitive/Types/UTxOSpec.hs (hunk 2 above)
Comment (HeinrichApfelmus) [reply lib/core/test/unit/Cardano/Wallet/Primitive/Types/UTxOSpec.hs]: Not necessarily, for two reasons:

* We not only want to check that `DeltaUTxO` is a monoid, but also that this monoid is compatible with `apply`. The latter property is more important, and actually implies that the monoid laws hold (assuming that `apply` is the only "observable"). In other words, the real test is whether `apply` is a monoid morphism.
* When checking the associativity law `(a <> b) <> c = a <> (b <> c)`, the values `a`,`b`, and `c` are generated randomly. But with the standard generators, it is exceedingly unlikely that the values contain overlapping `TxIn`. In this case, the `Semigroup` instance would be indistinguishable from the *wrong* instance
  ```
  instance Semigroup DeltaUTxO where
      db <> da = DeltaUTxO
          { excluded = excluded da <> excluded db -- wrong
          , received = received da <> received db -- nope
          }
  ```
    (I didn't try this, but it looks like an interesting exercise.)
    Instead of fiddling with generators to generate overlapping values, I might as well write down the complete Venn diagram; this relies on the assumption that `DeltaUTxO` is implemented only in terms of `Set` and `Map` and thus can be analyzed in terms of a Venn diagram.

File: lib/core/src/Cardano/Wallet/Primitive/Model.hs [hunk 3]
```diff
@@ -555,20 +583,32 @@ applyOurTxToUTxO
...
+applyOurTxToUTxO !slotNo !blockHeight !s !tx !u0 =
     let
         -- The next UTxO state (apply a state transition) (e.g. remove
         -- transaction outputs we've spent)
-        ourNextUTxO =
-            spendTx tx prevUTxO
-            <> UTxO.filterByAddress (ours s) (utxoFromTx tx)
+        (du, u) = (du21 <> du10, u2)
+
+        (du10, u1)   = spendTxD tx u0
```
Comment (paweljakubas) [inline lib/core/src/Cardano/Wallet/Primitive/Model.hs]: Maybe it would be good to add comment why du21 and du10 has this naming

File: lib/core/test/unit/Cardano/Wallet/Primitive/Types/UTxOSpec.hs [hunk 4]
```diff
@@ -78,6 +93,27 @@ spec =
...
 
+prop_deltaUTxO_semigroup_apply :: Property
+prop_deltaUTxO_semigroup_apply = 
+        delta2 `apply` (delta1 `apply` utxo0)
+    ===
+        (delta2 <> delta1) `apply` utxo0
+  where
+    -- As the functions in 'DeltaUTxO' are implemented in terms of set
+    -- operations, we only have to test a Venn diagram using few elements.
+    utxo0  = mkUTxOs ["a0","a1","a2"]
```
Comment (paweljakubas) [inline lib/core/test/unit/Cardano/Wallet/Primitive/Types/UTxOSpec.hs:141]: I like this manual example as it shows the case. One case. But what about also using proper arbitrary instances showing that deltaUTxO ops works properly for different overlapping of deltas? There are many possible cases here, right?

**Review by paweljakubas — APPROVED** [review]
lgtm! Maybe better generation of delta overlapping could be useful here

File: lib/core/test/unit/Cardano/Wallet/Primitive/Types/UTxOSpec.hs (hunk 4 above)
Comment (HeinrichApfelmus) [reply lib/core/test/unit/Cardano/Wallet/Primitive/Types/UTxOSpec.hs:141]: The beauty of elementary statements about sets is that they can be decided by looking at truth tables, which in turn corresponds to one universal example ("Venn diagram").

For example, in order to show that the equality `(A ∩ B) ∪ C = (A ∪ C) ∩ (B ∪ C)` holds for all finite sets A,B,C, it is sufficient to show that it holds for the specific choice `A = { "100", "110", "101", "111"}`, `B = { "010", "110", "011", "111"}`, and `C = { "001", "101", "011", "111"}`. Even though the elements like "010" seem very specific, they stand for an entire set of elements; here, "010" stands for all elements that are contained in `B`, but not in `A` or `C`.

The associativity of `DeltaUTxO` is a statement of this form, assuming that all operations are implemented in terms of set-theoretic operations like `Set.intersect`, which they essentially are. Hence, we can decide its validity with a single, universal example — the one in the test. This example can be visualize as follows:

venn diagram

In this sense, "a2" represents an entire set of outputs (suggested by the colored dots), namely the set of outputs contained in `utxo0`, not spent by `delta1`, but spent by `delta2`. Likewise, "a0" represents the set of outputs contained in `utxo0`, not spent by `delta1`, and not spent by `delta2`. Due to the "no double-spending" constraint, not all cases are possible; the diagram shows all the possible cases.

EDIT: I have also added a comment in the source code.

File: lib/core/src/Cardano/Wallet/Primitive/Model.hs (hunk 3 above)
Comment (HeinrichApfelmus) [reply lib/core/src/Cardano/Wallet/Primitive/Model.hs]: I have added a comment

**Comment by HeinrichApfelmus** [comment]
bors merge

**Comment by HeinrichApfelmus** [comment]
bors merge
